jax/_src/pallas/primitives.py

```diff
@@ -407,34 +413,42 @@ def _swap_discharge_rule(in_avals, out_avals, *args_flat, args_tree, **_):
   if len(indexers) > 1:
     raise NotImplementedError("Only one indexer supported in discharge rule.")
   idx = indexers[0]
-  if all((isinstance(s, Slice) or not s.shape) for s in idx.indices):
-    # TODO(b/329733289): support strided load/store in interpret mode.
-    for s in idx.indices:
-      if isinstance(s, Slice) and s.stride > 1:
-        raise NotImplementedError("Unimplemented stride support.")
+  if all((isinstance(s, Slice) or not s.shape) for s in idx.indices) or all(not isinstance(s, Slice) for s in idx.indices):
     indices = idx.indices
-    scalar_dims = [
-        i
-        for i, s in enumerate(indices)
-        if not isinstance(s, Slice) and not s.shape
-    ]
-    slice_starts = [s.start if isinstance(s, Slice) else s for s in indices]
-    slice_sizes = tuple(s.size if isinstance(s, Slice) else 1 for s in indices)
-    out = lax.dynamic_slice(ref, slice_starts, slice_sizes=slice_sizes)
-    out = jnp.squeeze(out, scalar_dims)
-    if mask is not None:
-      out_ = out
-      out = jnp.where(mask, out, val)
-      val = jnp.where(mask, val, out_)
-    val = jnp.expand_dims(val, scalar_dims)
-    x_new = lax.dynamic_update_slice(ref, val, start_indices=slice_starts)
-  elif all(not isinstance(s, Slice) for s in idx.indices):
-    out = ref[idx.indices]
+    if all((isinstance(s, Slice) or not s.shape) for s in idx.indices):
+      scalar_dims = [
+          i
+          for i, s in enumerate(indices)
+          if not isinstance(s, Slice) and not s.shape
+      ]
+      val = jnp.expand_dims(val, scalar_dims)
+      if (mask is not None):
+        mask = jnp.expand_dims(mask, scalar_dims)
+      slice_starts = [s.start if isinstance(s, Slice) else s for s in indices]
+      slice_sizes = tuple(s.size if isinstance(s, Slice) else 1 for s in indices)
+      slice_strides = tuple(s.stride if isinstance(s, Slice) else 1 for s in indices)
+      indices = [start + lax.iota(jnp.int32, size) * stride for (start, size, stride)
+                              in zip(slice_starts, slice_sizes, slice_strides)]
+      indices = tuple(jnp.meshgrid(*indices, indexing='ij'))
+
+    if (mask is not None):
+      # masked indexs do not occur, we simulate by using first element in array as dummy
+      is_first_element = (jnp.stack(indices) == 0).all(0)
```

Comment:
Same question here: why do we need to add special case handling to swap using the first element? Could we not solely rely on a jnp.where(mask, ...) to control which values get swapped as done in the original code?

Reply:
To mirror device behaviour, Ref is not accessed at masked indices. The best jittable approximation I found was to divert masked indices to index at the first element, this leads to masked indexing not occurring   which is often OOB indexing, while the first element should never be OOB.

Reply:
Dealt with in other comments